activesupport/test/message_verifier_test.rb

```diff
@@ -16,18 +16,133 @@ def load(value)
   end
 
   def setup
-    @verifier = ActiveSupport::MessageVerifier.new("Hey, I'm a secret!")
+    @secret = "Hey, I'm a secret!"
+    @verifier = ActiveSupport::MessageVerifier.new(@secret)
+    @payload = 'data'
+    @options = { expires_at: Time.local(2022), for: 'test' }
+    @claims = ActiveSupport::Claims.new(payload: @payload, **@options)
+  end
+
+  def test_valid_message
+    header, claims, digest = @verifier.generate(@payload, @options).split(".")
+    assert_not @verifier.valid_message?(nil)
+    assert_not @verifier.valid_message?("")
+    assert_not @verifier.valid_message?("\xff") # invalid encoding
+    assert_not @verifier.valid_message?("#{header.reverse}.#{claims}.#{digest}")
+    assert_not @verifier.valid_message?("#{header}.#{claims}.#{digest.reverse}")
+    assert_not @verifier.valid_message?("#{header}.#{claims.reverse}.#{digest}")
+    assert_not @verifier.valid_message?("purejunk")
+    assert_not @verifier.valid_message?("..")
+    assert_not @verifier.valid_message?("pure.junk.data")
+  end
+
+  def test_simple_round_tripping
+    message = @verifier.generate(@payload, @options)
+    assert_equal @payload, @verifier.verified(message, for: 'test')
+    assert_equal @payload, @verifier.verify(message, for: 'test')
+  end
+
+  def raise_invalid_signature_when_value_is_nil
```

Comment:
Here's a doubt:

* **Case 1: Previous ```MessageVerifer```**
Accepts ```nil``` as an argument in generate and returns a signed message. On calling ```verify``` on that signed_message, it raises ```ActiveSupport::MessageVerifier::InvalidSignature``` even when no tampering is done.
* **Case 2: New ```MessageVerifier```** 
I have implemented it same as previous ```MessageVerifier```.

However, I don't think ```generate``` should accept ```nil``` as a value at all. It doesn't make sense to sign nothing. What do you think @kaspth ?